std/range/package.d

```diff
@@ -9213,3 +9213,138 @@ if (is(typeof(fun) == void) || isSomeFunction!fun)
 
     auto r = [1, 2, 3, 4].tee!func1.tee!func2;
 }
+
+/**
+Extend the length of the input range $(D r) by padding out the start of the
+range with the element $(D e). The element $(D e) must be of a common type with
+the element type of the range $(D r) as defined by $(D std.traits.CommonType).
+
+Params:
+    r = an input range with a length
+    e = element to pad the range with
+    n = the length to pad to
+
+Return:
+    a range containing the elements of the original range with the extra padding
+*/
+auto padLeft(R, E)(R r, E e, size_t n) if (
+    isInputRange!R &&
+    (hasLength!R && !is(CommonType!(ElementType!R, E) == void)) ||
+    (isNarrowString!R && !is(CommonType!(R, E) == void)))
+{
+    static if (isNarrowString!R)
+    {
+        import std.algorithm.iteration : joiner;
+        return e.repeat(n > r.length ? n - r.length : 0).joiner.chain(r);
+    }
+    else
+    {
+        return e.repeat(n > r.length ? n - r.length : 0).chain(r);
+    }
+}
+
+///
+@safe pure unittest
+{
+    import std.algorithm : equal;
+
+    assert([1, 2, 3, 4].padLeft(0, 6).equal([0, 0, 1, 2, 3, 4]));
+    assert([1, 2, 3, 4].padLeft(0, 4).equal([1, 2, 3, 4]));
+
+    assert("abc".padLeft("_", 6).equal("___abc"));
+}
+
+@safe pure nothrow unittest
+{
+    import std.internal.test.dummyrange;
+    import std.algorithm : equal;
+    import std.meta : AliasSeq;
+
+    alias LengthDummyRanges = AliasSeq!(
+        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Input),
+        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Forward),
+        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Bidirectional),
+        DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Random),
+        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Input),
+        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Forward),
+        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Bidirectional),
+        DummyRange!(ReturnBy.Value, Length.Yes, RangeType.Random)
+    );
+
+    foreach(Range; LengthDummyRanges)
+    {
+        Range r;
+        assert(r
+            .padLeft(0, 12)
+            .equal([0, 0, 1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 9U, 10U])
+        );
+    }
+}
+
+/**
+Extend the length of the input range $(D r) by padding out the end of the range
+with the element $(D e). The element $(D e) must be of a common type with the
+element type of the range $(D r) as defined by $(D std.traits.CommonType).
+
+Params:
+    r = an input range with a length
+    e = element to pad the range with
+    n = the length to pad to
+
+Return:
+    a range containing the elements of the original range with the extra padding
+*/
+auto padRight(R, E)(R r, E e, size_t n) if (
+    isInputRange!R &&
+    (hasLength!R && !is(CommonType!(ElementType!R, E) == void)) ||
```

Comment:
The `padRight` version that doesn't require length seemed to be much of the point of the issue report. Will you work on adding that to this PR?